kpet/data.py

```diff
@@ -489,33 +485,49 @@ def convert_pattern(pattern):
                     matches.add(set_name)
             if matches == sets:
-                return set()
+                return None
             return matches
 
-        for suite in self.suites:
-            suite_sets = set()
-            for regex in suite.sets or set():
-                match = set(filter(regex.fullmatch, self.sets.keys()))
-                if match == set():
-                    raise Invalid(error.format("suite", regex.pattern))
-                suite_sets |= match
-            suite_sets |= convert_pattern(suite.pattern)
-            suite.sets = suite_sets
+        # This function is to just simplify the logic of combining
+        # the converted sets(from the old format) and sets from the new format
+        def combine_sets(first, second):
+            if first is None:
+                return second
+            if second is None:
+                return first
+            return first | second
 
-            for case in suite.cases:
-                case_sets = set()
-                for regex in case.sets or set():
+        for suite in self.suites:
+            if suite.sets is not None:
+                suite_sets = set()
+                for regex in suite.sets:
                     match = set(filter(regex.fullmatch, self.sets.keys()))
                     if match == set():
-                        raise Invalid(error.format("case", regex.pattern))
-                    case_sets |= match
-                case_sets |= convert_pattern(case.pattern)
-                # If the case doesn't specify any of it's own sets, we inherit
-                # the suite's sets
-                if case_sets == set():
-                    case_sets = suite.sets
-                case.sets = case_sets
-                if case.sets - suite.sets:
+                        raise Invalid(error.format("suite", regex.pattern))
+                    suite_sets |= match
+            else:
+                suite_sets = None
+
+            converted = convert_pattern(suite.pattern)
+            suite.sets = combine_sets(suite_sets, converted)
+
+            for case in suite.cases:
+                if case.sets is not None:
+                    case_sets = set()
+                    for regex in case.sets or set():
+                        match = set(filter(regex.fullmatch, self.sets.keys()))
+                        if match == set():
+                            raise Invalid(error.format("case", regex.pattern))
+                        case_sets |= match
+                else:
+                    case_sets = None
+                converted = convert_pattern(case.pattern)
+                case.sets = combine_sets(case_sets, converted)
+                if case.sets is None:
+                    # If the case doesn't specify any of it's own sets,
+                    # we inherit the suite's sets
+                    case.sets = suite.sets
+                if suite.sets is not None and case.sets - suite.sets:
```

Comment:
I think we still need the check for none, In the comment above we're checking for it because there might not be any case.sets specified in the yaml file. This check is there because when a suite / case doesn't "care" for it's sets (can also be just a pattern that doesn't mention sets) we set the .sets property to None as a wildcard, thus this None check is if we have used the wildcard.

Reply:
^^ nevermind, you're right